### ISSUE TITLE
Ability to find all training sample losses

### ISSUE THREAD

**Issue by enn-nafnlaus** [axolotl-ai-cloud/axolotl#901]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/OpenAccess-AI-Collective/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/OpenAccess-AI-Collective/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

When it comes to training data, the old adage is that quality matters more than quantity. The ability to filter out high-loss training samples from a dataset would enable one to use only the best, most  generalizable training samples from their dataset. This is of particular relevance when the training dataset is LLM-generated, and thus likely to contain a number of bad, non-representative generated samples.

### ✔️ Solution

Add some way (such as a command-line arg) to print or save the loss of each sample.  Probably the easiest way to filter would be to save the training data as a json with an extra parameter ("loss" or "eval_loss") appended to each sample - but simple printing of losses alongside the first X characters of the sample text would also suffice.

A sample commandline arg might be "--evalfile <filename>"

An ideal time to conduct this evaluation would be at training completion. A person who wanted an evaluation immediately could set epochs to 0.

### ❓ Alternatives

Honestly, the only alternative I can think of to try to mimic this is to try to train one sample at a time with a near-zero LR, but that's obviously a non-starter, given axolotl startup times.

I looked into trying to find the logging statement that prints this sort of output during training:

{'loss': 1.5473, 'learning_rate': 4.991435657859846e-06, 'epoch': 0.33}
{'loss': 1.5046, 'learning_rate': 4.991426650290067e-06, 'epoch': 0.33}
{'loss': 1.3384, 'learning_rate': 4.991417637994046e-06, 'epoch': 0.33}

... but couldn't find it, and regardless, even if I could find it, it would be nontrivial to associate it with a specific original training sample

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.
